crates/ra_syntax/src/grammar/type_params.rs

```diff
@@ -104,22 +104,32 @@ pub(super) fn opt_where_clause(p: &mut Parser) {
     }
     let m = p.start();
     p.bump();
-    loop {
-        if !(paths::is_path_start(p)
-            || p.current() == LIFETIME
-            || p.current() == FOR_KW
-            || p.current() == L_ANGLE)
-        {
-            break;
```

Comment:
Is it necessary to break if it is not a lifetime or type here?

Reply:
Yep, break is pretty important here for the error recovery. For the code like

```
fn foo()
where

struct Foo {}
```

we should end where clause when we see `struct` keyword, and that's what this `break` achieves.  Checking for `p.current() == L_CURLY || p.current() == SEMI || p.current() == EQ` will work for valid code correctly, but would be suboptimal for syntactically invalid code, which we need to handle as well.

I think `is_where_clause_end` should look like `!(p.at(LIFETIME) || p.at_ts(TYPE_FIRST) && !p.at(IMPL_KW))`. `TYPE_FIRST` is a set of tokens which can began a type.